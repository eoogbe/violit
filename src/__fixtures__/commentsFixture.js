/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as THREADS from './threadsFixture';
import * as USER from './userFixture';

export const comment1 = {
  id: 'comment1',
  dateCreated: '2019-10-09T10:11:12.123Z',
  dateDeleted: '2019-10-09T12:11:12.123Z',
  text: 'Sed do eiusmod tempor incididunt ut labore et dolore.',
  threadSlug: THREADS.thread1.slug,
  author: USER.username,
};

export const comment2 = {
  id: 'comment2',
  dateCreated: '2019-10-08T07:08:09.101Z',
  dateDeleted: '2019-10-08T07:09:09.101Z',
  text: 'Ut enim ad minim veniam.',
  threadSlug: THREADS.thread1.slug,
  author: USER.username,
};

export const comment3 = {
  id: 'comment3',
  dateCreated: '2019-09-30T01:02:03.456Z',
  dateDeleted: '2019-10-03T01:02:03.456Z',
  text: 'Quis nostrud exercitation ullamco laboris nisi ut.',
  threadSlug: THREADS.thread1.slug,
  author: USER.username,
};

export const createInput = {
  id: 'comment1',
  text: comment1.text,
  threadSlug: THREADS.thread1.slug,
};

export const normalized = {
  comment1: {
    id: 'comment1',
    dateCreated: comment1.dateCreated,
    dateDeleted: null,
    text: comment1.text,
    threadSlug: comment1.threadSlug,
    author: comment1.author,
  },
  comment2: {
    id: 'comment2',
    dateCreated: comment2.dateCreated,
    dateDeleted: comment2.dateDeleted,
    text: comment2.text,
    threadSlug: comment2.threadSlug,
    author: comment2.author,
  },
  comment3: {
    id: 'comment3',
    dateCreated: comment3.dateCreated,
    dateDeleted: null,
    text: comment3.text,
    threadSlug: comment3.threadSlug,
    author: comment3.author,
  },
};

export const unnormalized = [
  normalized.comment1,
  normalized.comment2,
  normalized.comment3,
];

export const api = [
  {
    href: 'http://localhost:4000/api/comments/comment1',
    data: [
      { name: 'id', value: 'comment1' },
      { name: 'dateCreated', value: comment1.dateCreated },
      { name: 'dateDeleted', value: '' },
      { name: 'text', value: comment1.text },
      { name: 'author', value: comment1.author },
      { name: 'threadSlug', value: THREADS.thread1.slug },
    ],
    links: [
      {
        rel: 'deleteAction',
        href: 'http://localhost:4000/api/comments/comment1/delete',
      },
      { rel: 'about', href: THREADS.thread1.url },
    ],
  },
  {
    href: 'http://localhost:4000/api/comments/comment2',
    data: [
      { name: 'id', value: 'comment2' },
      { name: 'dateCreated', value: comment2.dateCreated },
      { name: 'dateDeleted', value: comment2.dateDeleted },
      { name: 'text', value: comment2.text },
      { name: 'author', value: comment2.author },
      { name: 'threadSlug', value: THREADS.thread1.slug },
    ],
    links: [
      { rel: 'about', href: THREADS.thread1.url },
    ],
  },
  {
    href: 'http://localhost:4000/api/comments/comment3',
    data: [
      { name: 'id', value: 'comment3' },
      { name: 'dateCreated', value: comment3.dateCreated },
      { name: 'dateDeleted', value: '' },
      { name: 'text', value: comment3.text },
      { name: 'author', value: comment3.author },
      { name: 'threadSlug', value: THREADS.thread1.slug },
    ],
    links: [
      {
        rel: 'deleteAction',
        href: 'http://localhost:4000/api/comments/comment3/delete',
      },
      { rel: 'about', href: THREADS.thread1.url },
    ],
  },
];

export const deleteCommentApi = [
  {
    href: 'http://localhost:4000/api/comments/comment1/delete',
    data: [
      { name: 'actionStatus', value: 'completedActionStatus' },
      { name: 'endTime', value: comment1.dateDeleted },
    ],
    links: [
      {
        rel: 'object',
        href: 'http://localhost:4000/api/comments/comment1',
      },
      { rel: 'targetCollection', href: THREADS.thread1.commentsUrl },
    ],
  },
];

export const commentsFetch = {
  url: THREADS.thread1.commentsUrl,
  method: 'GET',
  status: 200,
  body: {
    items: api,
  },
};

export const createCommentFetch = {
  url: THREADS.thread1.commentsUrl,
  method: 'POST',
  status: 201,
  body: {
    items: [api[0]],
  },
};

export const canCreateCommentFetch = {
  url: THREADS.thread1.commentsUrl,
  method: 'OPTIONS',
  status: 204,
  headers: {
    Allow: 'GET,HEAD,OPTIONS,POST',
  },
};

export const cannotCreateCommentFetch = {
  url: THREADS.thread1.commentsUrl,
  method: 'OPTIONS',
  status: 204,
  headers: {
    Allow: 'GET,HEAD,OPTIONS',
  },
};

export const deleteCommentFetch = {
  url: 'http://localhost:4000/api/comments/comment1/delete',
  method: 'POST',
  status: 200,
  body: {
    items: deleteCommentApi,
  },
};

export const db = [
  {
    id: 'comment1',
    dateCreated: comment1.dateCreated,
    dateDeleted: null,
    content: comment1.text,
    userId: 'user1',
    threadId: 'thread1',
  },
  {
    id: 'comment2',
    dateCreated: comment2.dateCreated,
    dateDeleted: comment2.dateDeleted,
    content: comment2.text,
    userId: 'user1',
    threadId: 'thread1',
  },
  {
    id: 'comment3',
    dateCreated: comment3.dateCreated,
    dateDeleted: null,
    content: comment3.text,
    userId: 'user1',
    threadId: 'thread1',
  },
];
