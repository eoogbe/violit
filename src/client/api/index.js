/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { load, loadAll } from './load';
export { default as InvalidResponseError } from './InvalidResponseError';
export { default as checkExists } from './checkExists';
export { default as checkPermissions } from './checkPermissions';
export { default as href } from './href';
export { default as sendCreate } from './sendCreate';
export { default as sendDelete } from './sendDelete';
export { default as sendPost } from './sendPost';
