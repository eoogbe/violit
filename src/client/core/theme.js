/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export default {
  primaryColor: '#7F00FF',

  gray200: '#EEEEEE',
  gray300: '#E0E0E0',
  gray500: '#9E9E9E',
  gray700: '#616161',

  red200: '#EF9A9A',
  red500: '#F44336',

  lightBlue700: '#0288D1',

  black: '#000000',
  white: '#FFFFFF',

  spacingSm: '0.25em',
  spacingMd: '0.5em',
  spacingLg: '1em',

  borderRadius: '2px',
};
