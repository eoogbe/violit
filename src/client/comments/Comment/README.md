```jsx
import * as COMMENTS from '__fixtures__/commentsFixture';

<Comment
  id="comment1"
  author={COMMENTS.comment1.author}
  dateCreated={COMMENTS.comment1.dateCreated}
  text={COMMENTS.comment1.text}
  canDelete
/>
```

When deleted:

```jsx
import * as COMMENTS from '__fixtures__/commentsFixture';

<Comment
  id="comment1"
  author={COMMENTS.comment2.author}
  dateCreated={COMMENTS.comment2.dateCreated}
  dateDeleted={COMMENTS.comment2.dateDeleted}
/>
```
