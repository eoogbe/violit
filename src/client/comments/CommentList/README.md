```jsx
import * as COMMENTS from '__fixtures__/commentsFixture';
import * as THREADS from '__fixtures__/threadsFixture';

mockFetch([
  COMMENTS.commentsFetch,
]);

<CommentList threadSlug={THREADS.thread1.slug} />
```

When loading (refresh page if loaded):

```jsx
import * as THREADS from '__fixtures__/threadsFixture';
import { wait } from 'testUtils/wait';

window.fetch = () => wait(1000000);

<CommentList threadSlug={THREADS.thread1.slug} />
```

When error:

```jsx
import * as THREADS from '__fixtures__/threadsFixture';

window.fetch = () => Promise.reject();

<CommentList threadSlug={THREADS.thread1.slug} />
```

Without comments:

```jsx
import * as THREADS from '__fixtures__/threadsFixture';

mockFetch([
  {
    url: THREADS.thread1.commentsUrl,
    method: 'GET',
    status: 200,
    body: {
      items: [],
    },
  },
]);

<CommentList threadSlug={THREADS.thread1.slug} />
```
