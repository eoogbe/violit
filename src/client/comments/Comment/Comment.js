/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { ArticleFooter, ArticleHeader } from 'client/article';
import { Markdown } from 'client/core';
import * as api from '../api';

const StyledComment = styled.li`
  font-size: 1rem;
  margin: ${({ theme }) => theme.spacingLg} 0;
  padding: 0 ${({ theme }) => theme.spacingMd};
  border-left: 2px solid ${({ theme }) => theme.gray300};
  list-style: none;
`;

const NoTextParagraph = styled.p`
  font-size: 1rem;
  margin: ${({ theme }) => theme.spacingMd} 0;
  color: ${({ theme }) => theme.gray700};
`;

NoTextParagraph.displayName = 'NoTextParagraph';

/** A comment reply. */
const Comment = (props) => {
  const { id, author, dateCreated, text, onDeleteError } = props;
  const [dateDeleted, setDateDeleted] = useState(props.dateDeleted);
  const [canDelete, setCanDelete] = useState(props.canDelete);

  const deleteComment = useCallback(async () => {
    setCanDelete(false);
    setDateDeleted(new Date().toISOString());

    try {
      await api.sendDeleteComment(id);
    } catch (err) {
      if (onDeleteError) {
        onDeleteError(err);
      }
    }
  }, [id, onDeleteError]);

  return (
    <StyledComment
      itemScope
      itemType="http://schema.org/Comment"
      itemProp="comment"
    >
      <article>
        <ArticleHeader author={author} dateCreated={dateCreated} />
        <div itemProp="text">
          {!dateDeleted && <Markdown source={text} />}
          {dateDeleted && <NoTextParagraph>[deleted]</NoTextParagraph>}
        </div>
        {canDelete &&
          <ArticleFooter
            contentId={`comment-${id}`}
            deleteContent={deleteComment}
          />}
      </article>
    </StyledComment>
  );
};

Comment.propTypes = {
  /** The unique id of the comment. */
  id: PropTypes.string.isRequired,
  /** The username of the person who wrote the comment. */
  author: PropTypes.string.isRequired,
  /** The date the comment was posted. */
  dateCreated: PropTypes.string.isRequired,
  /** The date the comment was deleted. */
  dateDeleted: PropTypes.string,
  /** The text content of the comment. */
  text: PropTypes.string,
  /** `true` if the current user can delete the comment. */
  canDelete: PropTypes.bool,
  /** Called when comment deletion fails. */
  onDeleteError: PropTypes.func,
};

export default Comment;
