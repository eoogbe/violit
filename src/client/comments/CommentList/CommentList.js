/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { forwardRef, useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import Loading from 'react-loading';
import styled from 'styled-components';
import { EmptyListParagraph } from 'client/article';
import { Alert, theme } from 'client/core';
import { useLoad } from 'client/loading';
import Comment from '../Comment';
import * as api from '../api';

const StyledError = styled.div`
  font-size: 1rem;
  position: relative;
  height: 3.5em;
`;

const StyledCommentList = styled.ul`
  margin: 0;
  padding: 0;
  list-style: none;
`;

const StyledLoading = styled(Loading)`
  font-size: 1rem;
  margin: ${({ theme }) => theme.spacingMd} auto;
`;
/** Renders a single comment item. */
const renderComment = ({ onDeleteError }) => (comment) => (
  <Comment key={comment.id} {...comment} onDeleteError={onDeleteError} />
);

/** The list of comments for a thread. */
const CommentList = forwardRef(({ threadSlug, additionalComments }, ref) => {
  const [deleteError, setDeleteError] = useState(false);

  const loadComments = useCallback(() => {
    return api.loadComments(threadSlug);
  }, [threadSlug]);

  const [
    loading,
    loadError,
    loadedComments,
  ] = useLoad(loadComments);

  if (loadError) {
    return (
      <StyledError>
        <Alert action="loading the comments" />
      </StyledError>
    );
  }

  if (deleteError) {
    return (
      <StyledError>
        <Alert action="deleting the comment" />
      </StyledError>
    );
  }

  const comments =
    [].concat(additionalComments, loadedComments).filter(Boolean);
  if (comments.length) {
    return (
      <StyledCommentList ref={ref}>
        {comments.map(renderComment({
          onDeleteError: () => setDeleteError(true),
        }))}
      </StyledCommentList>
    );
  }

  if (loading) {
    return (
      <StyledLoading type="spinningBubbles" color={theme.primaryColor} />
    );
  }

  return <EmptyListParagraph>No comments</EmptyListParagraph>;
});

CommentList.propTypes = {
  /** The slug of the parent thread. */
  threadSlug: PropTypes.string.isRequired,
  /** Comments to prepend to the loaded comments. */
  additionalComments: PropTypes.arrayOf(PropTypes.shape({
    /** The unique id of the comment. */
    id: PropTypes.string.isRequired,
    /** The username of the person who wrote the comment. */
    author: PropTypes.string.isRequired,
    /** The date the comment was posted. */
    dateCreated: PropTypes.string.isRequired,
    /** The date the comment was deleted. */
    dateDeleted: PropTypes.string,
    /** The text content of the comment. */
    text: PropTypes.string,
    /** `true` if the current user can delete the comment. */
    canDelete: PropTypes.bool,
  })),
  /** A ref to the list. */
  ref: PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
};

export default CommentList;
