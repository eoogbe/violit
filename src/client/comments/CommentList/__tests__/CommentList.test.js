/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { createRef } from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import TestRenderer from 'react-test-renderer';
import * as COMMENTS from '__fixtures__/commentsFixture';
import * as THREADS from '__fixtures__/threadsFixture';
import mockFetch from 'testUtils/mockFetch';
import { wait } from 'testUtils/wait';
import { Alert } from 'client/core';
import Comment from '../../Comment';
import CommentList from '../CommentList';

const NOW = 1572566400000;

const ADDITIONAL_COMMENTS = [
  {
    id: 'comment1',
    dateCreated: COMMENTS.comment1.dateCreated,
    dateDeleted: null,
    text: COMMENTS.comment1.text,
    threadSlug: COMMENTS.comment1.threadSlug,
    author: COMMENTS.comment1.author,
    canDelete: true,
  },
];

it('renders without crashing', async () => {
  Date.now = jest.fn().mockReturnValue(NOW);
  mockFetch([
    {
      url: THREADS.thread1.commentsUrl,
      method: 'GET',
      status: 200,
      body: {
        items: [COMMENTS.api[1], COMMENTS.api[2]],
      },
    },
  ]);

  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <CommentList
        threadSlug={THREADS.thread1.slug}
        additionalComments={ADDITIONAL_COMMENTS}
      />
    );

    await wait();
  });

  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders when loading', () => {
  global.fetch = jest.fn().mockReturnValue(wait(5));

  let renderer;
  TestRenderer.act(() => {
    renderer = TestRenderer.create(
      <CommentList threadSlug={THREADS.thread1.slug} />
    );
  });

  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders with load error', async () => {
  global.fetch = jest.fn().mockRejectedValue(null);

  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <CommentList threadSlug={THREADS.thread1.slug} />
    );

    await wait();
  });

  expect(renderer.toJSON()).toMatchSnapshot();
});

it('renders when empty', async () => {
  mockFetch([
    {
      url: THREADS.thread1.commentsUrl,
      method: 'GET',
      status: 200,
      body: {
        items: [],
      },
    },
  ]);

  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <CommentList threadSlug={THREADS.thread1.slug} />
    );

    await wait();
  });

  expect(renderer.toJSON()).toMatchSnapshot();
});

it('forwards the ref', async () => {
  mockFetch([
    COMMENTS.commentsFetch,
  ]);
  const ref = createRef();

  await act(async () => {
    mount(<CommentList threadSlug={THREADS.thread1.slug} ref={ref} />);

    await wait();
  });

  expect(ref.current).toBeTruthy();
});

// TOOD(#68): Figure out how to manipulate an enzyme wrapper for an element
// loaded asynchronously
it.skip('with delete error shows alert', async () => {
  mockFetch([
    COMMENTS.commentsFetch,
    {
      url: 'http://localhost:4000/api/comments/comment1/delete',
      method: 'POST',
      status: 500,
    },
  ]);

  const wrapper = mount(<CommentList threadSlug={THREADS.thread1.slug} />);

  await wait();

  expect(wrapper.find(Alert)).toHaveLength(0);

  wrapper.find(Comment).first().find('OverflowToggle').simulate('click');
  wrapper
    .find(Comment)
    .first()
    .find('MenuItemButton')
    .first()
    .simulate('click');

  expect(wrapper.find(Alert).text()).toContain('deleting the comment');
});
