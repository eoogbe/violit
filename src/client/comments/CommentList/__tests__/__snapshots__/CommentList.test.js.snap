// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when empty 1`] = `
<p
  className="sc-jTzLTM dHyGjY"
>
  No comments
</p>
`;

exports[`renders when loading 1`] = `
<div
  className="sc-kGXeez fqyRQz"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\">
  <circle cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(45 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.125s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(90 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.25s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(135 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.375s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(180 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.5s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(225 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.625s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(270 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.75s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(315 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.875s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
  <circle transform=\\"rotate(180 16 16)\\" cx=\\"16\\" cy=\\"3\\" r=\\"0\\">
    <animate attributeName=\\"r\\" values=\\"0;3;0;0\\" dur=\\"1s\\" repeatCount=\\"indefinite\\" begin=\\"0.5s\\" keySplines=\\"0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8\\" calcMode=\\"spline\\" />
  </circle>
</svg>
",
    }
  }
  style={
    Object {
      "fill": "#7F00FF",
      "height": 64,
      "width": 64,
    }
  }
/>
`;

exports[`renders with load error 1`] = `
<div
  className="sc-chPdSV iHoWTf"
>
  <div
    aria-live="assertive"
    className="sc-bdVaJa dXBAhf"
    role="alert"
  >
    <p
      className="sc-htpNat bhpHJD"
    >
      Oops! Something went wrong with 
      loading the comments
      !Â 
      <button
        className="sc-bxivhb ErYiX"
        onClick={[Function]}
        type="button"
      >
        Reload page
      </button>
    </p>
    <button
      className="sc-bwzfXH haUZkm"
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="1.25em"
        size="1.25em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        title="Dismiss alert"
        viewBox="0 0 24 24"
        width="1.25em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<ul
  className="sc-kgoBCf dbbukP"
>
  <li
    className="sc-fjdhpX hXRbNK"
    itemProp="comment"
    itemScope={true}
    itemType="http://schema.org/Comment"
  >
    <article>
      <header
        className="sc-gqjmRU dBUhVe"
      >
        <span
          className="sc-VigVT djDVvN"
          itemProp="author"
        >
          fred-astaire
        </span>
        <time
          dateTime="2019-10-09T10:11:12.123Z"
          itemProp="dateCreated"
          title="2019-10-09T10:11:12.123Z"
        >
          3 weeks ago
        </time>
      </header>
      <div
        itemProp="text"
      >
        <div
          className="sc-bZQynM kESxfG"
        >
          <p>
            Sed do eiusmod tempor incididunt ut labore et dolore.
          </p>
        </div>
      </div>
      <footer
        className="sc-dnqmqq fWoffO"
      >
        <nav
          className="sc-iwsKbI eoiWGh"
        >
          <button
            aria-controls="overflow-menu-comment-comment1"
            aria-expanded={false}
            aria-haspopup="true"
            className="sc-gZMcBi kfjqfz"
            onClick={[Function]}
            type="button"
          >
            <svg
              fill="currentColor"
              focusable={false}
              height="1.5em"
              size="1.5em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              title="More actions"
              viewBox="0 0 24 24"
              width="1.5em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
          </button>
        </nav>
      </footer>
    </article>
  </li>
  <li
    className="sc-fjdhpX hXRbNK"
    itemProp="comment"
    itemScope={true}
    itemType="http://schema.org/Comment"
  >
    <article>
      <header
        className="sc-gqjmRU dBUhVe"
      >
        <span
          className="sc-VigVT djDVvN"
          itemProp="author"
        >
          fred-astaire
        </span>
        <time
          dateTime="2019-10-08T07:08:09.101Z"
          itemProp="dateCreated"
          title="2019-10-08T07:08:09.101Z"
        >
          3 weeks ago
        </time>
      </header>
      <div
        itemProp="text"
      >
        <p
          className="sc-jzJRlG cnQzZH"
        >
          [deleted]
        </p>
      </div>
    </article>
  </li>
  <li
    className="sc-fjdhpX hXRbNK"
    itemProp="comment"
    itemScope={true}
    itemType="http://schema.org/Comment"
  >
    <article>
      <header
        className="sc-gqjmRU dBUhVe"
      >
        <span
          className="sc-VigVT djDVvN"
          itemProp="author"
        >
          fred-astaire
        </span>
        <time
          dateTime="2019-09-30T01:02:03.456Z"
          itemProp="dateCreated"
          title="2019-09-30T01:02:03.456Z"
        >
          1 month ago
        </time>
      </header>
      <div
        itemProp="text"
      >
        <div
          className="sc-bZQynM kESxfG"
        >
          <p>
            Quis nostrud exercitation ullamco laboris nisi ut.
          </p>
        </div>
      </div>
      <footer
        className="sc-dnqmqq fWoffO"
      >
        <nav
          className="sc-iwsKbI eoiWGh"
        >
          <button
            aria-controls="overflow-menu-comment-comment3"
            aria-expanded={false}
            aria-haspopup="true"
            className="sc-gZMcBi kfjqfz"
            onClick={[Function]}
            type="button"
          >
            <svg
              fill="currentColor"
              focusable={false}
              height="1.5em"
              size="1.5em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              title="More actions"
              viewBox="0 0 24 24"
              width="1.5em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
          </button>
        </nav>
      </footer>
    </article>
  </li>
</ul>
`;
